File: app/src/main/java/org/beiwe/app/ui/user/MainMenuActivity.java
package org.beiwe.app.ui.user;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import org.beiwe.app.R;
import org.beiwe.app.session.SessionActivity;
import org.beiwe.app.storage.PersistentData;
import org.beiwe.app.survey.SurveyActivity;
import org.beiwe.app.ui.utils.SurveyNotifications;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

/**The main menu activity of the app. Currently displays 4 buttons - Audio Recording, Graph, Call Clinician, and Sign out.
 * @author Dor Samet */
public class MainMenuActivity extends SessionActivity {
	//extends a SessionActivity

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main_menu);

		Button callClinicianButton = (Button) findViewById(R.id.main_menu_call_clinician);
		if(PersistentData.getCallClinicianButtonEnabled()) {
			callClinicianButton.setText(PersistentData.getCallClinicianButtonText());
		}
		else {
			callClinicianButton.setVisibility(View.GONE);
		}

		ArrayList<String> permSurveyIds = new ArrayList<String>();
		for (String surveyId : PersistentData.getSurveyIds() ){
			try {
				JSONObject surveySettings = new JSONObject(PersistentData.getSurveySettings(surveyId));
				if (surveySettings.getBoolean("always_available")) {
					permSurveyIds.add(surveyId);
				}
			}
			catch (JSONException e) {e.printStackTrace();}
		}
		if (permSurveyIds.size() !=0 ) {
			for (int i = 0; i < permSurveyIds.size(); i++) {
				Button button = (Button) findViewById(getResources().getIdentifier("permSurvey" + i, "id", this.getPackageName()));
				if (PersistentData.getSurveyType(permSurveyIds.get(i)).equals("audio_survey")){
					button.setText(R.string.permaaudiosurvey);
				}
				button.setTag(R.string.permasurvey, permSurveyIds.get(i));
				button.setVisibility(View.VISIBLE);
			}
		}
	}
	
	/*#########################################################################
	############################## Buttons ####################################
	#########################################################################*/
	public void displaySurvey(View view) {
		Intent activityIntent;
		String surveyId = (String) view.getTag(R.string.permasurvey);

		if (PersistentData.getSurveyType(surveyId).equals("audio_survey")){
			activityIntent = new Intent(getApplicationContext(), SurveyNotifications.getAudioSurveyClass(surveyId));
		} else {
			activityIntent = new Intent(getApplicationContext(), SurveyActivity.class);
		}
		activityIntent.setAction( getApplicationContext().getString(R.string.start_tracking_survey) );
		activityIntent.putExtra("surveyId", surveyId);
		activityIntent.setFlags( Intent.FLAG_ACTIVITY_CLEAR_TOP );
		startActivity(activityIntent);
	}

//	public void graphResults (View v) { startActivity( new Intent(getApplicationContext(), GraphActivity.class) ); }
}

File: app/src/main/java/org/beiwe/app/ui/user/MainMenuActivity.kt
package org.beiwe.app.ui.user

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import kotlinx.android.synthetic.main.activity_main_menu.main_menu_call_clinician
import org.beiwe.app.R
import org.beiwe.app.session.SessionActivity
import org.beiwe.app.storage.PersistentData
import org.beiwe.app.survey.SurveyActivity
import org.beiwe.app.ui.utils.SurveyNotifications
import org.json.JSONException
import org.json.JSONObject

/**The main menu activity of the app. Currently displays 4 buttons - Audio Recording, Graph, Call Clinician, and Sign out.
 * @author Dor Samet
 */
class MainMenuActivity : SessionActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main_menu)

        if (PersistentData.getCallClinicianButtonEnabled())
            main_menu_call_clinician.text = PersistentData.getCallClinicianButtonText()
        else
            main_menu_call_clinician.visibility = View.GONE
    }

    fun setupSurveyList() {
        // get the active and always available surveys
        val surveyIds = ArrayList<String>()
        for (surveyId in PersistentData.getSurveyIds())
            try {
                val surveySettings = JSONObject(PersistentData.getSurveySettings(surveyId))
                val is_active = SurveyNotifications.isNotificationActive(applicationContext, surveyId)
                if (surveySettings.getBoolean("always_available") || is_active)
                    surveyIds.add(surveyId)
            } catch (e: JSONException) {
                e.printStackTrace()
            }

        var button_count = 0
        for (i in surveyIds.indices) {
            button_count = i
            var surveyName = PersistentData.getSurveyName(surveyIds[i])
            val surveyType = PersistentData.getSurveyType(surveyIds[i])
            val button = findViewById<View>(resources.getIdentifier(
                    "permSurvey$i", "id", this.packageName)) as Button

            // set textAllCaps and surveyName if there is no survey name
            if (surveyName == "") {
                button.isAllCaps = true
                surveyName = if (surveyType == "audio_survey")
                    getString(R.string.permaaudiosurvey)
                else
                    getString(R.string.perm_survey)
            } else
                button.isAllCaps = false

            // add emoji to survey name if it is an audio survey
            if (surveyType == "audio_survey")
                surveyName = "$surveyName 🎙"

            // button.text = "$surveyName ($survey_tag)"
            button.text = surveyName
            button.setTag(R.string.permasurvey, surveyIds[i])
            button.visibility = View.VISIBLE

            // there are 16 buttons, so we only need to iterate 0 to 16
            if (i >= 16)
                break
        }

        // iterate over every unused button, disable button and GONE it to fix scroll bugs
        for (i in button_count + 1..16) {
            val button = findViewById<View>(resources.getIdentifier(
                    "permSurvey$i", "id", this.packageName)) as Button
            button.visibility = View.GONE
        }
    }

    // run an task every 5 seconds to refresh the available survey list
    override fun onResume() {
        super.onResume()
        setupSurveyList()
    }


    /*#########################################################################
	############################## Buttons ####################################
	#########################################################################*/
    fun displaySurvey(view: View) {
        val activityIntent: Intent
        val surveyId = view.getTag(R.string.permasurvey) as String
        activityIntent = if (PersistentData.getSurveyType(surveyId) == "audio_survey") {
            Intent(applicationContext, SurveyNotifications.getAudioSurveyClass(surveyId))
        } else {
            Intent(applicationContext, SurveyActivity::class.java)
        }
        activityIntent.action = applicationContext.getString(R.string.start_tracking_survey)
        activityIntent.putExtra("surveyId", surveyId)
        activityIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        startActivity(activityIntent)
    }
}